### PR TITLE
HPSPOVERQ functionality

### PR DESCRIPTION
**Pull request by sarojaduality**
- HPSPOVERQLEVELED still doesn't work for higher multiplicative depth.

### CODE REVIEW

**Review by yspolyakov — APPROVED** [review]

File: src/pke/lib/scheme/bfvrns/bfvrns-cryptoparameters.cpp [hunk 1]
```diff
@@ -261,7 +271,7 @@ void CryptoParametersBFVRNS::PrecomputeCRTTables(
       }
       m_paramsQlRl[0] =
           std::make_shared<ILDCRTParams<BigInteger>>(2 * n, moduliQR, rootsQR);
-    } else if (multTech == HPSPOVERQLEVELED) {
+    } else if (multTech == HPSPOVERQLEVELED || multTech == HPSPOVERQ) {
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/bfvrns/bfvrns-cryptoparameters.cpp:274]: Should we have "else" for this if statement? for multTech BEHZ or invalid. The same question applies to some code in this file below.

Comment (sarojaduality) [reply src/pke/lib/scheme/bfvrns/bfvrns-cryptoparameters.cpp:274]: No, we don't need any of these precomputations for BEHZ. There is separate code in the function for the BEHZ precomputations, but they all have different names.

File: src/pke/lib/scheme/bfvrns/bfvrns-cryptoparameters.cpp [hunk 2]
```diff
@@ -192,28 +192,35 @@ void CryptoParametersBFVRNS::PrecomputeCRTTables(
...
+    BigInteger tmpModulusQ = modulusQ;
+    
+    if (multTech == HPSPOVERQLEVELED || multTech == HPSPOVERQ) {
       m_QlHatInvModq.resize(sizeQ);
       m_QlHatInvModqPrecon.resize(sizeQ);
       m_QlHatModr.resize(sizeQ);
 
       for (size_t l = 0; l < sizeQ; l++) {
-        if (l > 0) modulusQ = modulusQ / BigInteger(moduliQ[sizeQ - l]);
+        if (l > 0) tmpModulusQ = tmpModulusQ / BigInteger(moduliQ[sizeQ - l]);
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/scheme/bfvrns/bfvrns-cryptoparameters.cpp:204]: They are always within the bounds.

File: src/pke/include/schemerns/rns-cryptoparameters.h [hunk 3]
```diff
@@ -638,6 +638,19 @@ class CryptoParametersRNS : public CryptoParametersRLWE<DCRTPoly> {
   // BFVrns : Mult : ExpandCRTBasis
   /////////////////////////////////////
 
+  const std::shared_ptr<ILDCRTParams<BigInteger>> GetParamsQl(usint l = 0) const {
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/include/schemerns/rns-cryptoparameters.h:641]: Noted as a future issue to resolve. Not sure about how it affects performance right now.

File: src/pke/include/scheme/bfvrns/gen-cryptocontext-bfvrns-internal.h [hunk 4]
```diff
@@ -95,6 +95,16 @@ typename ContextGeneratorType::ContextType genCryptoContextBFVRNSInternal(const
     // for BFV scheme noise scale is always set to 1
     params->SetNoiseScale(1);
 
+	uint32_t numLargeDigits = parameters.GetNumLargeDigits();
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/include/scheme/bfvrns/gen-cryptocontext-bfvrns-internal.h:98]: Fixed in next PR.

**Pull request merged**